plat/kvm/x86/setup.c

```diff
@@ -1,456 +1,295 @@
-	 * pagetable64.S).
+	/* TODO: Until we generate the boot page table at compile time, we
+	 * manually add an untyped unmap region to the boot info to force an
+	 * unmapping of the 1:1 mapping after the kernel image before mapping
+	 * only the necessary parts.
 	 */
-	offset = mr[0]->start - PAGE_LARGE_ALIGN_UP(mr[0]->start);
-	start  = PAGE_LARGE_ALIGN_UP(mr[0]->start);
-	len    = PAGE_LARGE_ALIGN_DOWN(mr[0]->len - offset);
-
-	rc = ukplat_page_unmap(&kernel_pt, start,
-			       (PLATFORM_MAX_MEM_ADDR - start) >> PAGE_SHIFT,
-			       PAGE_FLAG_KEEP_FRAMES);
-	if (unlikely(rc))
-		goto EXIT_FATAL;
+	rc = ukplat_memregion_list_insert(&bi->mrds,
+		&(struct ukplat_memregion_desc){
+			.vbase = PAGE_ALIGN_UP(__END),
+			.pbase = 0,
+			.len   = PLATFORM_MAX_MEM_ADDR - PAGE_ALIGN_UP(__END),
+			.type  = 0,
+			.flags = UKPLAT_MEMRF_UNMAP,
+		});
+	if (unlikely(rc < 0))
+		return rc;
+
+	return paging_init();
+}
+#else /* CONFIG_HAVE_PAGING */
+static int mem_init(struct ukplat_bootinfo *bi)
+{
+	struct ukplat_memregion_desc *mrdp;
+	int i;
 
-	/* Add remaining physical memory that has not been added to the heaps
-	 * previously. Also map regions marked as RESERVED 1:1 because this
-	 * will include memory-mapped ACPI and APIC tables and registers. For
-	 * now, we map these ranges as read-only.
+	/* The static boot page table maps only the first 4 GiB. Remove all
+	 * free memory regions above this limit so we won't use them for the
+	 * heap. Start from the tail as the memory list is ordered by address.
+	 * We can stop at the first area that is completely in the mapped area.
 	 */
-	for (offset = 0; offset < mi->mmap_length;
-	     offset += m->size + sizeof(m->size)) {
-		m = (void *)(__uptr)(mi->mmap_addr + offset);
-
-		if (m->addr <= PLATFORM_MEM_START)
-			continue;
-
-		if (m->type == MULTIBOOT_MEMORY_RESERVED) {
-			vaddr  = paddr = PAGE_ALIGN_DOWN(m->addr);
-			frames = DIV_ROUND_UP(m->len, PAGE_SIZE);
-
-			rc = ukplat_page_map(&kernel_pt, vaddr, paddr, frames,
-					     PAGE_ATTR_PROT_READ, 0);
-			if (unlikely(rc))
-				goto EXIT_FATAL;
-		} else if (m->type == MULTIBOOT_MEMORY_AVAILABLE) {
-			rc = ukplat_pt_add_mem(&kernel_pt, m->addr, m->len);
-			if (unlikely(rc))
-				goto EXIT_FATAL;
+	for (i = (int)bi->mrds.count - 1; i >= 0; i--) {
+		ukplat_memregion_get(i, &mrdp);
+		if (mrdp->vbase >= PLATFORM_MAX_MEM_ADDR) {
+			/* Region is outside the mapped area */
+			uk_pr_info("Memory %012lx-%012lx outside mapped area\n",
+				   mrdp->vbase, mrdp->vbase + mrdp->len);
+
+			if (mrdp->type == UKPLAT_MEMRT_FREE)
+				ukplat_memregion_list_delete(&bi->mrds, i);
+		} else if (mrdp->vbase + mrdp->len > PLATFORM_MAX_MEM_ADDR) {
+			/* Region overlaps with unmapped area */
+			uk_pr_info("Memory %012lx-%012lx outside mapped area\n",
+				   PLATFORM_MAX_MEM_ADDR,
+				   mrdp->vbase + mrdp->len);
+
+			if (mrdp->type == UKPLAT_MEMRT_FREE)
+				mrdp->len -= (mrdp->vbase + mrdp->len) -
+						PLATFORM_MAX_MEM_ADDR;
```

Comment:
Perhaps another niptick, but since you are assuming the list of memory descriptors is already ordered descendingly you could `break;` here as well, or drop the `else if` entirely:
 ```
         } else {
			if (mrdp->vbase + mrdp->len > PLATFORM_MAX_MEM_ADDR) {
				/* Region overlaps with unmapped area */
				uk_pr_info("Memory %012lx-%012lx outside mapped area\n",
					   PLATFORM_MAX_MEM_ADDR,
				           mrdp->vbase + mrdp->len);

				if (mrdp->type == UKPLAT_MEMRT_FREE)
					mrdp->len -= (mrdp->vbase + mrdp->len) -
						     PLATFORM_MAX_MEM_ADDR;
			}

			break;
		}
```
though that would add another level of indentation.

Reply:
Yes, makes sense. I added the `break` to avoid the additional level of indentation. It is perhaps also a bit easier to grasp.